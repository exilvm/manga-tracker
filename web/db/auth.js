const pool = require('.');
const LRU = require("lru-cache");
const crypto = require('crypto');
const { bruteforce } = require('./../utils/ratelimits');

const sessionDebug = require('debug')('session-debug');
const authInfo = require('debug')('auth-info');

const userCache = new LRU(({
                max: 50,
                maxAge: 86400000, // 1 day in ms
                noDisposeOnSet: true,
                updateAgeOnGet: true,
}));

function generateAuthToken(uid, user_uuid, cb) {
    crypto.randomBytes(32+9, (err, buf) => {
    if (err) {
        return cb(err, false);
    }

    const token = buf.toString('base64', 0, 33);
    const lookup = buf.toString('base64', 33);

    const sql = `INSERT INTO auth_tokens (user_id, hashed_token, expires_at, lookup) VALUES ($1, encode(digest($2, 'sha256'), 'hex'), $3, $4)`;
    const age = 2592e+6; // 30 days
    pool.query(sql, [uid, token, new Date(Date.now() + age), lookup])
        .then(() => cb(null, `${lookup};${token};${Buffer.from(user_uuid).toString('base64')}`))
        .catch(err => cb(err, false));
    });
}
module.exports.generateAuthToken = generateAuthToken;

function regenerateAuthToken(uid, lookup, user_uuid, cb) {
    crypto.randomBytes(32, (err, buf) => {
    if (err) {
        return cb(err, false);
    }

    const token = buf.toString('base64');

    const sql = `UPDATE auth_tokens SET hashed_token=encode(digest($3, 'sha256'), 'hex') WHERE user_id=$1 AND lookup=$2 RETURNING expires_at`;
    pool.query(sql, [uid, lookup, token])
        .then(res => {
            if (res.rowCount === 0 || res.rows.length === 0) return cb(null, false);
            cb(null, `${lookup};${token};${Buffer.from(user_uuid).toString('base64')}`, res.rows[0].expires_at)
        })
        .catch(err => cb(err, false))
    });
}

module.exports.authenticate = function (req, email, password, cb) {
    if (password.length > 72) return cb(null, false);

    let sql = `SELECT user_id, username, user_uuid, theme FROM users WHERE email=$1 AND pwhash=crypt($2, pwhash)`
    pool.query(sql, [email, password])
        .then(res => {
        if (res.rowCount === 0) {
            return cb(null, false);
        }
        const row = res.rows[0];

        function setUser(row, token) {
            // Try to regen session
            req.session.regenerate((err)=>{
                if (err) {
                    console.error(err);
                    req.session.user_id = undefined;
                    return cb(err, false);
                }
                req.session.user_id = row.user_id;
                userCache.set(row.user_id, {
                    user_id: row.user_id,
                    username: row.username,
                    uuid: row.user_uuid,
                    theme: row.theme
                });
                return cb(null, token);
            });
}

        if (req.body.rememberme !== "on") {
            return setUser(row, true);
        }

        generateAuthToken(row.user_id, row.user_uuid, (err, token) => {
            if (err) return cb(err, false);
            setUser(row, token)
        })
        })
        .catch(err => {
            console.error(err);
            cb(err, false);
        });
}

function getUser(uid, cb) {
    if (!uid) return cb(null, null);

    let user = userCache.get(uid);
    if (user) return cb(user, null);

    const sql = `SELECT username, user_uuid, theme FROM users WHERE user_id=$1`;
    pool.query(sql, [uid])
         .then(res => {
            if (res.rowCount === 0) return cb(null, null);
            const row = res.rows[0];
            const val = {
                username: row.username,
                uuid: row.user_uuid,
                user_id: uid,
                theme: row.theme
            }
            userCache.set(uid, val);
            cb(val, null);
        })
        .catch(err => {
            console.error(err);
            cb(null, err);
        });
}
module.exports.getUser = getUser;

module.exports.requiresUser = function(req, res, next) {
    getUser(req.session.user_id, (user, err) => {
        req.user = user
        next(err);
    });
}

module.exports.checkAuth =  function(app) {
    return function (req, res, next) {
        // We don't need to check authentication for resources generated by next.js
        if (req.session.user_id || !req.cookies.auth || req.originalUrl.startsWith('/_next')) return next();
        bruteforce.prevent(req, res, () => {
            // TODO Maybe fix the race condition when 2 requests are done back to back
            //  with the same token from the same ip. Or maybe by some other criteria

            authInfo('Checking auth from db for', req.cookies.auth);
            const sql = `SELECT u.user_id, u.username, u.user_uuid, u.theme FROM auth_tokens INNER JOIN users u on u.user_id=auth_tokens.user_id 
                         WHERE expires_at > NOW() AND user_uuid=$1 AND 
                               lookup=$2 AND hashed_token=encode(digest($3, 'sha256'), 'hex')`;

            /*
            Try to find the remember me token.
            If found associate current session with user and regenerate session id (this is important)
            If something fails or user isn't found we remove possible user id from session and continue
             */
            let [lookup, token, uuid] = req.cookies.auth.split(';', 3);
            if (!uuid) {
                res.clearCookie('auth');
                return next();
            }

            uuid = Buffer.from(uuid, 'base64').toString('ascii');
            if (uuid.length < 32) {
                res.clearCookie('auth');
                return next();
            }

            pool.query(sql, [uuid, lookup, token])
                .then(sqlRes => {
                    if (sqlRes.rowCount === 0) {
                        sessionDebug('Session not found. Clearing cookie')
                        res.clearCookie('auth');
                        req.session.user_id = undefined;

                        const checkLookup = `SELECT u.user_id FROM auth_tokens 
                                                              INNER JOIN users u ON auth_tokens.user_id = u.user_id 
                                             WHERE user_uuid=$1 AND lookup=$2`;

                        pool.query(checkLookup, [uuid, lookup])
                            .then(res2 => {
                                if (res2.rowCount === 0) return next();
                                // TODO Display warning
                                clearUserAuthTokens(res2.rows[0].user_id, () => {
                                    app.sessionStore.clearUserSessions(res2.rows[0].user_id, () => {
                                        sessionDebug("Invalid auth token found for user. Sessions cleared");
                                        next()
                                    })
                                })
                            })
                            .catch(next)
                        return;
                    }

                    const row = sqlRes.rows[0];
                    userCache.set(row.user_id, {
                        user_id: row.user_id,
                        username: row.username,
                        uuid: row.user_uuid,
                        theme: row.theme
                    });
                    // Try to regen session
                    req.session.regenerate((err)=>{
                        if (err) {
                            req.session.user_id = undefined;
                            return next(err);
                        }
                        regenerateAuthToken(row.user_id, lookup, uuid, (err, token, expiresAt) => {
                            if (err || !token) {
                                console.error('Failed to regenerate/change token', err);
                                return next(err);
                            }
                            authInfo('regen', token)

                            req.session.user_id = row.user_id;
                            res.cookie('auth', token, {
                                httpOnly: true,
                                sameSite: 'strict',
                                expires: expiresAt
                            });
                            return next();
                        })
                    });
                })
                .catch(err => {
                    req.session.user_id = undefined;
                    if (err.code === '22P02') return next(err);
                    console.error(err);
                    next(err);
                })
        });
    }
}

function clearUserAuthToken(uid, auth, cb) {
    const [lookup, token] = auth.split(';', 3);

    const sql = `DELETE FROM auth_tokens WHERE user_id=$1 AND lookup=$2 AND hashed_token=encode(digest($3, 'sha256'), 'hex')`;
    pool.query(sql, [uid, lookup, token])
        .then(() => cb(null))
        .catch(err => cb(err));
}

module.exports.clearUserAuthToken = clearUserAuthToken;

function clearUserAuthTokens(uid, cb) {
    const sql = `DELETE FROM auth_tokens WHERE user_id=$1`;
    pool.query(sql, [uid])
        .then(() => cb(null))
        .catch(err => cb(err));
}

module.exports.clearUserAuthTokens = clearUserAuthTokens;

module.exports.modifyCacheUser = (uid, modifications) => {
    const user = userCache.get(uid);
    if (!user) return;
    userCache.set(uid, {...user, ...modifications});
}