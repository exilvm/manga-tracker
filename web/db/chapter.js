const db = require('.');

module.exports.getChapterReleases = (mangaId) => {
  const sql = `SELECT extract(EPOCH FROM date_trunc('day', release_date)) as "timestamp", CAST(count(release_date) as int) count 
               FROM chapters 
               WHERE manga_id=$1 GROUP BY 1 ORDER BY 1`;

  return db.query(sql, [mangaId])
    .then(res => Promise.resolve(res.rows));
};

module.exports.getChapters = (mangaId, limit, offset) => {
  const sql = `
    SELECT
        COUNT(*) as count,
        (
            SELECT json_agg(ch)
            FROM (
                SELECT
                    chapter_id,
                    title,
                    chapter_number,
                    EXTRACT(EPOCH from release_date) as release_date,
                    "group",
                    service_id,
                    chapter_identifier as chapter_url
                FROM chapters WHERE manga_id=$1
                ORDER BY chapter_number DESC, chapter_decimal DESC NULLS LAST
                LIMIT $2 ${offset ? 'OFFSET $3' : ''}
            ) as ch
        ) as chapters
    FROM chapters
    WHERE manga_id=$1
  `;

  const args = [mangaId, limit];
  if (offset) args.push(offset);

  return db.query(sql, args)
    .then(res => {
      const row = res.rows[0];
      if (!row) return Promise.resolve(null);

      return Promise.resolve({
        count: row.count,
        chapters: row.chapters,
      });
    });
};
