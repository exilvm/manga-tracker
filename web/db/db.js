const pool = require('.');

function getLatestReleases(service_id, manga_id, user_uuid) {
    const joins = [];
    const where = [];
    const args = [];
    let paramNumber = 1;
    if (user_uuid) {
        joins.push(`INNER JOIN user_follows uf ON c.manga_id = uf.manga_id AND (uf.service_id IS NULL OR c.service_id=uf.service_id) 
                    INNER JOIN users u ON u.user_id=uf.user_id`);
        where.push(`u.user_uuid=$${paramNumber}::uuid`);
        args.push(user_uuid);
        paramNumber++;
    }

    if (manga_id) {
        where.push(`c.manga_id=$${paramNumber}`);
        args.push(manga_id);
        paramNumber++;
    }

    if (service_id) {
        where.push(`c.service_id=$${paramNumber}`);
        args.push(service_id);
        paramNumber++;
    }

    const sql = `
        WITH chapters_filtered AS (
            SELECT chapter_id, title, chapter_number, chapter_decimal, release_date, chapter_identifier, c.service_id, c.manga_id 
            FROM chapters c ${joins.join(' ')}
            ${where.length > 0 ? 'WHERE ' + where.join(' AND ') : ''}
            ORDER BY release_date DESC, chapter_number DESC
        )
        SELECT c.chapter_id, m.title as manga_title, m.manga_id, m.release_interval, c.title, c.chapter_number, c.chapter_decimal, c.release_date, c.chapter_identifier, s.service_name, s.chapter_url_format, s.url
        FROM chapters_filtered c INNER JOIN manga m on c.manga_id = m.manga_id INNER JOIN services s on c.service_id = s.service_id 
        WHERE c.release_date > NOW() - INTERVAL '1 hour'
        UNION 
              (SELECT c.chapter_id, m.title as manga_title, m.manga_id, m.release_interval, c.title, c.chapter_number, c.chapter_decimal, c.release_date, c.chapter_identifier, s.service_name, s.chapter_url_format, s.url
              FROM chapters_filtered c INNER JOIN manga m on c.manga_id = m.manga_id INNER JOIN services s on c.service_id = s.service_id
              LIMIT 30)
        ORDER BY release_date DESC, chapter_number DESC`;

    return pool.query(sql, args);
}

function getUserFollows(user_id, manga_id) {
    const sql = `SELECT service_id FROM user_follows WHERE user_id=$1 AND manga_id=$2`;
    return pool.query(sql, [user_id, manga_id]);
}

module.exports = { getLatestReleases, getUserFollows }