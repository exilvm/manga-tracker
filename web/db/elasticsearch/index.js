const { Client } = require('@elastic/elasticsearch');

const client = new Client({
  node: process.env.ELASTIC_NODE,
  auth: {
    username: process.env.ELASTIC_USERNAME,
    password: process.env.ELASTIC_PASSWORD,
  },
});

module.exports = client;
