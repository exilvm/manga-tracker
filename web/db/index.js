const isTest = process.env.NODE_ENV === 'test';

const pgp = require('pg-promise')({
  noLocking: isTest,
  capSQL: true,
  pgFormatting: true,
});

// Will cause warnings when in development mode
const db = pgp({
  host: process.env.DB_HOST,
  user: process.env.DB_USER,
  database: isTest ?
    process.env.DB_NAME_TEST || process.env.DB_NAME :
    process.env.DB_NAME,
  port: process.env.DB_PORT,
  password: process.env.PGPASSWORD,
  max: 5,
  idleTimeoutMillis: 30000,
  connectionTimeoutMillis: 2000,
});

function query(sql, args) {
  return db.query(sql, args);
}

async function end() {
  await pgp.end();
}

module.exports = {
  query,
  db,
  pgp,
  end,
};
