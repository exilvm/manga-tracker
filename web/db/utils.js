const format = require('pg-format');
const {
  NUMERIC_VALUE_OUT_OF_RANGE,
  INVALID_TEXT_REPRESENTATION,
} = require('pg-error-constants');
const dblog = require('debug')('db');

function generateEqualsColumns(o, availableColumns) {
  const cols = new Set(availableColumns);
  const sqlValues = [];
  const columns = [];
  const args = [];

  Object.keys(o)
    .filter(k => o[k] !== undefined && cols.has(k))
    .forEach((k, idx) => {
      sqlValues.push(`%I=$${idx+1}`);
      columns.push(k);
      args.push(o[k]);
    });

  const sqlCols = format(sqlValues.join(','), ...columns);

  return { sqlCols, args };
}
module.exports.generateEqualsColumns = generateEqualsColumns;

function handleError(err, res) {
  if (err.code === INVALID_TEXT_REPRESENTATION) {
    dblog(err.message);
    res.status(400).json({ error: 'Invalid data type given' });
    return;
  }

  if (err.code === NUMERIC_VALUE_OUT_OF_RANGE) {
    res.status(400).json({ error: 'Number value out of range' });
    return;
  }

  console.error(err);
  res.status(500).json({ error: 'Internal server error' });
}
module.exports.handleError = handleError;
