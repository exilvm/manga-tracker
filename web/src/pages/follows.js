/* eslint-disable global-require,camelcase */
import React from 'react';
import withError from '../utils/withError';
import Follows from '../views/Follows';

const MangaPage = function MangaPage(props) {
  const {
    follows,
  } = props;

  return <Follows follows={follows} />;
};


export async function getServerSideProps({ req }) {
  const { getFollows } = require('../../db/manga');
  let error;
  let follows;
  try {
    follows = await getFollows(req.user?.user_id);
  } catch (e) {
    error = e;
  }

  if (!follows || error) {
    return { props: { error: error?.status || 404 }};
  }

  return {
    props: {
      follows: follows,
    },
  };
}
export default withError(MangaPage);
