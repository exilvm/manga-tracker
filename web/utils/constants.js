const csrfMissing = 'Missing CSRF token';
const isDev = process.env.NODE_ENV !== 'production';
const isTest = process.env.NODE_ENV === 'test';


module.exports = {
  csrfMissing,
  isDev,
  isTest,
};
