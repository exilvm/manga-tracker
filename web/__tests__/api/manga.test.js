import request from 'supertest';
import { insertFollow } from '../../db/follows';
import { userUnauthenticated, userUnauthorized, mangaIdError } from '../constants';

import initServer from '../initServer';
import stopServer from '../stopServer';
import { adminUser, expectErrorMessage, normalUser, withUser } from '../utils';

let httpServer;

beforeAll(async () => {
  ({ httpServer } = await initServer());
});

afterAll(async () => {
  await stopServer(httpServer);
});

describe('PUT /api/user/follows', () => {
  it('Returns 403 without user authentication', async () => {
    await request(httpServer)
      .put(`/api/user/follows`)
      .expect(403)
      .expect(expectErrorMessage(userUnauthenticated));
  });

  it('Returns 400 with invalid manga id', async () => {
    const errorMessage = mangaIdError;

    await withUser(normalUser, async () => {
      await request(httpServer)
        .put(`/api/user/follows?manga_id=-1`)
        .expect(400)
        .expect(expectErrorMessage('-1', 'manga_id', errorMessage));

      await request(httpServer)
        .put(`/api/user/follows?manga_id=Infinity`)
        .expect(400)
        .expect(expectErrorMessage('Infinity', 'manga_id', errorMessage));

      await request(httpServer)
        .put(`/api/user/follows?manga_id=`)
        .expect(400)
        .expect(expectErrorMessage('', 'manga_id', errorMessage));

      await request(httpServer)
        .put(`/api/user/follows?manga_id=2147483648`)
        .expect(400)
        .expect(expectErrorMessage('Number value out of range'));
    });
  });

  it('Returns 400 with invalid service id', async () => {
    const errorMessage = 'Service id must be a positive integer';

    await withUser(normalUser, async () => {
      await request(httpServer)
        .put(`/api/user/follows?manga_id=1&service_id=abc`)
        .expect(400)
        .expect(expectErrorMessage('abc', 'service_id', errorMessage));

      await request(httpServer)
        .put(`/api/user/follows?manga_id=1&service_id=`)
        .expect(400)
        .expect(expectErrorMessage('', 'service_id', errorMessage));

      await request(httpServer)
        .put(`/api/user/follows?manga_id=1&service_id=-1`)
        .expect(400)
        .expect(expectErrorMessage('-1', 'service_id', errorMessage));

      await request(httpServer)
        .put(`/api/user/follows?manga_id=1&service_id=undefined`)
        .expect(400)
        .expect(expectErrorMessage('undefined', 'service_id', errorMessage));
    });
  });

  it('Returns 200 with valid data', async () => {
    await withUser(normalUser, async () => {
      await request(httpServer)
        .put(`/api/user/follows?manga_id=1&service_id=1`)
        .expect(200);

      await request(httpServer)
        .put(`/api/user/follows?manga_id=1`)
        .expect(200);
    });
  });
});

describe('DELETE /api/user/follows', () => {
  it('Returns 403 without user authentication', async () => {
    await request(httpServer)
      .delete(`/api/user/follows`)
      .expect(403)
      .expect(expectErrorMessage(userUnauthenticated));
  });

  it('Returns 400 with invalid manga id', async () => {
    const errorMessage = mangaIdError;

    await withUser(normalUser, async () => {
      await request(httpServer)
        .delete(`/api/user/follows?manga_id=-1`)
        .expect(400)
        .expect(expectErrorMessage('-1', 'manga_id', errorMessage));

      await request(httpServer)
        .delete(`/api/user/follows?manga_id=Infinity`)
        .expect(400)
        .expect(expectErrorMessage('Infinity', 'manga_id', errorMessage));

      await request(httpServer)
        .delete(`/api/user/follows?manga_id=`)
        .expect(400)
        .expect(expectErrorMessage('', 'manga_id', errorMessage));

      await request(httpServer)
        .delete(`/api/user/follows?manga_id=2147483648`)
        .expect(400)
        .expect(expectErrorMessage('Number value out of range'));
    });
  });

  it('Returns 400 with invalid service id', async () => {
    const errorMessage = 'Service id must be a positive integer';

    await withUser(normalUser, async () => {
      await request(httpServer)
        .delete(`/api/user/follows?manga_id=1&service_id=abc`)
        .expect(400)
        .expect(expectErrorMessage('abc', 'service_id', errorMessage));

      await request(httpServer)
        .delete(`/api/user/follows?manga_id=1&service_id=`)
        .expect(400)
        .expect(expectErrorMessage('', 'service_id', errorMessage));

      await request(httpServer)
        .delete(`/api/user/follows?manga_id=1&service_id=-1`)
        .expect(400)
        .expect(expectErrorMessage('-1', 'service_id', errorMessage));

      await request(httpServer)
        .delete(`/api/user/follows?manga_id=1&service_id=undefined`)
        .expect(400)
        .expect(expectErrorMessage('undefined', 'service_id', errorMessage));
    });
  });

  it('Returns 200 with valid input and 404 on non existent resource', async () => {
    await insertFollow(normalUser.user_id, 1, 1);
    await insertFollow(normalUser.user_id, 1, null);

    await withUser(normalUser, async () => {
      await request(httpServer)
        .delete(`/api/user/follows?manga_id=1&service_id=1`)
        .expect(200);

      // Make sure resource was deleted
      await request(httpServer)
        .delete(`/api/user/follows?manga_id=1&service_id=1`)
        .expect(404);

      await request(httpServer)
        .delete(`/api/user/follows?manga_id=1`)
        .expect(200);

      // Make sure resource was deleted
      await request(httpServer)
        .delete(`/api/user/follows?manga_id=1`)
        .expect(404);
    });
  });
});


describe('POST /api/manga/merge', () => {
  it('Returns 403 without user authentication', async () => {
    await request(httpServer)
      .post('/api/manga/merge')
      .expect(403)
      .expect(expectErrorMessage(userUnauthenticated));
  });

  it('Returns 401 without admin rights', async () => {
    await withUser(normalUser, async () => {
      await request(httpServer)
        .post('/api/manga/merge')
        .expect(401)
        .expect(expectErrorMessage(userUnauthorized));
    });
  });

  it('Returns 400 with invalid base id', async () => {
    await withUser(adminUser, async () => {
      await request(httpServer)
        .post('/api/manga/merge?base=-1')
        .expect(400)
        .expect(expectErrorMessage('-1', 'base'));

      await request(httpServer)
        .post('/api/manga/merge?base=NaN')
        .expect(400)
        .expect(expectErrorMessage('NaN', 'base'));

      await request(httpServer)
        .post('/api/manga/merge?base=abc')
        .expect(400)
        .expect(expectErrorMessage('abc', 'base'));
    });
  });

  it('Returns 400 with invalid to_merge id', async () => {
    await withUser(adminUser, async () => {
      await request(httpServer)
        .post('/api/manga/merge?to_merge=-1')
        .expect(400)
        .expect(expectErrorMessage('-1', 'to_merge'));

      await request(httpServer)
        .post('/api/manga/merge?to_merge=NaN')
        .expect(400)
        .expect(expectErrorMessage('NaN', 'to_merge'));

      await request(httpServer)
        .post('/api/manga/merge?to_merge=abc')
        .expect(400)
        .expect(expectErrorMessage('abc', 'to_merge'));
    });
  });

  it('Return 400 when base equals to_merge', async () => {
    await withUser(adminUser, async () => {
      await request(httpServer)
        .post('/api/manga/merge?to_merge=5&base=5')
        .expect(400)
        .expect(expectErrorMessage('Given ids are equal'));
    });
  });
});


describe('GET /api/manga/:manga_id', () => {
  it('returns 404 with non existent manga id', async () => {
    await request(httpServer)
      .get(`/api/manga/9999999`)
      .expect(404)
      .expect(expectErrorMessage('Manga not found'));
  });

  it('returns 404 with non positive integers', async () => {
    await request(httpServer)
      .get(`/api/manga/-1`)
      .expect(404);

    await request(httpServer)
      .get(`/api/manga/abc`)
      .expect(404);
  });

  const chapterError = 'Amount of chapters must be a positive integer';
  it('returns 400 with invalid chapter count', async () => {
    await request(httpServer)
      .get(`/api/manga/1?chapters=`)
      .expect(400)
      .expect(expectErrorMessage('', 'chapters', chapterError));

    await request(httpServer)
      .get(`/api/manga/1?chapters=abc`)
      .expect(400)
      .expect(expectErrorMessage('abc', 'chapters', chapterError));

    await request(httpServer)
      .get(`/api/manga/1?chapters=-1`)
      .expect(400)
      .expect(expectErrorMessage('-1', 'chapters', chapterError));
  });

  it('returns 200 with valid manga id and chapters', async () => {
    const checkManga = res => {
      expect(res.body).toBeObject();
      expect(res.body.manga).toBeTruthy();
    };

    await request(httpServer)
      .get(`/api/manga/1`)
      .expect(200)
      .expect(checkManga);

    await request(httpServer)
      .get(`/api/manga/1?chapters=5`)
      .expect(200)
      .expect(checkManga)
      .expect(res => expect(res.body.manga.chapters).toHaveLength(5));
  });
});


const getChapterCount = (body) => (body && body.chapters?.length) || 0;

describe('GET /api/manga/:manga_id/chapters', () => {
  const validUrl = '/api/manga/1/chapters';

  it('returns 404 with invalid manga id', async () => {
    await request(httpServer)
      .get(`/api/manga/9999999/chapters`)
      .expect(404);

    await request(httpServer)
      .get(`/api/manga/-1/chapters`)
      .expect(404);
  });

  it('returns 400 with an invalid limit', async () => {
    const errorMessage = 'Limit must be an integer between 0 and 200';
    await Promise.all([
      request(httpServer)
        .get(`${validUrl}?limit=invalid`)
        .expect('Content-Type', /json/)
        .expect(expectErrorMessage('invalid', 'limit', errorMessage))
        .expect(400),

      request(httpServer)
        .get(`${validUrl}?limit=NaN`)
        .expect('Content-Type', /json/)
        .expect(expectErrorMessage('NaN', 'limit', errorMessage))
        .expect(400),

      request(httpServer)
        .get(`${validUrl}?limit=Infinity`)
        .expect('Content-Type', /json/)
        .expect(expectErrorMessage('Infinity', 'limit', errorMessage))
        .expect(400),

      request(httpServer)
        .get(`${validUrl}?limit=-1`)
        .expect('Content-Type', /json/)
        .expect(expectErrorMessage('-1', 'limit', errorMessage))
        .expect(400),

      request(httpServer)
        .get(`${validUrl}?limit=300`)
        .expect('Content-Type', /json/)
        .expect(expectErrorMessage('300', 'limit', errorMessage))
        .expect(400),
    ]);
  });

  it('Returns chapters with valid limit', async () => {
    await Promise.all([
      request(httpServer)
        .get(`${validUrl}`)
        .expect('Content-Type', /json/)
        .expect(res => expect(getChapterCount(res.body)).toBeGreaterThan(0))
        .expect(200),

      request(httpServer)
        .get(`${validUrl}?limit=5`)
        .expect('Content-Type', /json/)
        .expect(res => expect(getChapterCount(res.body)).toEqual(5))
        .expect(200),
    ]);
  });

  it('Returns 400 with invalid offset', async () => {
    const errorMessage = 'Offset must be a positive integer';

    await Promise.all([
      request(httpServer)
        .get(`${validUrl}?offset=NaN`)
        .expect('Content-Type', /json/)
        .expect(expectErrorMessage('NaN', 'offset', errorMessage))
        .expect(400),

      request(httpServer)
        .get(`${validUrl}?offset=-1`)
        .expect('Content-Type', /json/)
        .expect(expectErrorMessage('-1', 'offset', errorMessage))
        .expect(400),
    ]);
  });

  it('Returns chapters with valid offset', async () => {
    await Promise.all([
      request(httpServer)
        .get(`${validUrl}?offset=3`)
        .expect('Content-Type', /json/)
        .expect(res => expect(getChapterCount(res.body)).toBeGreaterThan(0))
        .expect(200),
    ]);
  });
});
