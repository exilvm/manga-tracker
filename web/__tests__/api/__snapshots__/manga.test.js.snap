// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /api/manga/merge Returns 400 with invalid service id 1`] = `
Array [
  Object {
    "location": "query",
    "msg": "Invalid value",
    "param": "service",
    "value": "-1",
  },
]
`;

exports[`POST /api/manga/merge Returns 400 with invalid service id 2`] = `
Array [
  Object {
    "location": "query",
    "msg": "Invalid value",
    "param": "service",
    "value": "abc",
  },
]
`;
